src/qibo/backends/__init__.py

```diff
@@ -0,0 +1,113 @@
+from qibo import config
+from qibo.config import raise_error, log, warnings, BACKEND_NAME
+from qibo.backends.numpy import NumpyBackend
+from qibo.backends.tensorflow import TensorflowBackend
+
+
+CONSTRUCTED_BACKENDS = {}
+def construct_backend(name):
+    if name not in CONSTRUCTED_BACKENDS:
+        if name == "numpy":
+            CONSTRUCTED_BACKENDS["numpy"] = NumpyBackend()
+        elif name == "tensorflow":
+            CONSTRUCTED_BACKENDS["tensorflow"] = TensorflowBackend()
+        else:
+            raise_error(ValueError, "Unknown backend name {}.".format(name))
+    return CONSTRUCTED_BACKENDS.get(name)
+
+numpy_backend = construct_backend("numpy")
+numpy_matrices = numpy_backend.matrices
+
+if BACKEND_NAME == "tensorflow":
+    K = TensorflowBackend()
+else: # pragma: no cover
+    # CI uses tensorflow as default backend
+    K = NumpyBackend()
+
+def set_backend(backend="custom"):
+    """Sets backend used to implement gates.
+
+    Args:
+        backend (str): possible options are 'custom' for the gates that use
+            custom tensorflow operator and 'defaulteinsum' or 'matmuleinsum'
+            for the gates that use tensorflow primitives (``tf.einsum`` or
+            ``tf.matmul`` respectively).
+    """
+    if not config.ALLOW_SWITCHERS and backend != K.gates:
+        warnings.warn("Backend should not be changed after allocating gates.",
+                      category=RuntimeWarning)
+
+    gate_backend = backend.split("_")
```

Comment:
I am not sure to understand why there is a distinction between say `defaulteinsum` and `numpy_defaulteinsum`, if I understand correctly, it falls back to ['defaulteinsum', 'matmuleinsum', 'custom'], right?

Another question, concerning `custom` with the `numpy` backend, I think the `set_backend` should prevent this operation.

Reply:
> I am not sure to understand why there is a distinction between say `defaulteinsum` and `numpy_defaulteinsum`, if I understand correctly, it falls back to ['defaulteinsum', 'matmuleinsum', 'custom'], right?

The difference is that `custom`, `defaulteinsum` and `matmuleinsum` all use Tensorflow, while `numpy_defaulteinsum` and `numpy_matmuleinsum` use only numpy and can be used in an environment without Tensorflow installed.

> Another question, concerning `custom` with the `numpy` backend, I think the `set_backend` should prevent this operation.

In the most recent push, `set_backend` raises an error if one attempts to use the custom backend without tensorflow. However, in principle (I have not tested), it should be possible to use the tensorflow custom operators with the numpy backend. In this case all ops excluding the custom operators will be done in numpy. However there may be performance deterioration (particularly for small qubit numbers) because the custom operators will automatically cast the numpy arrays to `tf.Tensor` when a custom operator is called, so I am not sure if we would like to have this option.